### ISSUE TITLE
Package Pirat

### ISSUE THREAD

**Issue by samWieczorek** [Bioconductor/Contributions#3308]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/prostarproteomics/Pirat

Confirm the following by editing each check box to '[x]'

- [ x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [ x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x ] I understand Bioconductor [Package Naming Policy][9] and acknowledge
  Bioconductor may retain use of package name.

- [ x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x ] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x ] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.
  
- [x ] I am familiar with the [Bioconductor code of conduct][7] and 
  agree to abide by it.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [ x] The 'devel' branch for new packages and features.
- [x ] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x ] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For questions/help about the submission process, including questions about
the output of the automatic reports generated by the SPB (Single Package
Builder), please use the #package-submission channel of our Community Slack.
Follow the link on the home page of the [Bioconductor website][8] to sign up.

[1]: https://contributions.bioconductor.org/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/
[7]: https://bioconductor.org/about/code-of-conduct/
[8]: https://bioconductor.org/
[9]: https://bioconductor.org/developers/package-submission/#naming

**Comment by vjcitn** [comment]
Thanks for this contribution.  Please see https://bioconductor.org/packages/3.19/bioc/html/basilisk.html for the preferred approach to declaring and using python dependencies.  I see that you have LBFGS optimization in python source code, and a second python file that uses it via `import python.LBFGS`.  It would be good to indicate why a scipy optimizer won't do here.  You can use reticulate within basilisk to get your sources using `source_python`.  This would enable you to remove the miniconda installation and the manipulation of RETICULATE_PYTHON from your package.  Happy to discuss further.

**Comment by lshep** [comment]
may we expect updates to address the above concern?

**Comment by samWieczorek** [comment]
hi

thank you for getting back to me because I hadn't seen your first message and I've just read it. I'll see how I can modify Pirat to simplify installation. Indeed, if I could do without installing miniconda and avoid having to handle RETICULATE_PYTHON, I'd prefer that.  I'll keep you posted on this concern.

I was just about to write to you because I have a question about using the set.seed function, which generates a warning when I execute the BiocCheck::BiocCheck('new-package'=TRUE) function.

I use the set.seed function to be able to reproduce exactly the results obtained which are published in an article. It's important for me to be able to continue using it. Is there a solution for keeping it and remaining compliant with Bioconductor?

Thank you for your help

**Comment by lshep** [comment]
A set seed can be used in examples and vignettes just make sure it is done outside any internal package code. it should be set visibly, before a function is called.

**Comment by samWieczorek** [comment]
Following your advice, I modified the Pirat code and now use the basilisk package. It's much simpler. All I have to do now is fix the problem with the set.seed() function

**Comment by lshep** [comment]
Please let us know when it is addressed and we will re-evaluate moving this forward in the process.

**Comment by samWieczorek** [comment]
Hi, I have fixed this warning with the set.seed() function. Please let me know if i have to do anything else ?

**Comment by lshep** [comment]
Please fix the ERROR in the build report before the package will move into review.

**Comment by samWieczorek** [comment]
I have posted a question on the Bioc mailing list but have no answer yet.
Here is the question: 

As you adviced me, I use the package basilisk

I have followed the instructions of the basilisk package to build mine. Expecially, there are two files named ‘configure’ and ‘configure.win’ which are necessary to install any package which import basilisk.

However, I still have the following build error :
ERROR: 'configure' exists but is not executable -- see the 'R Installation and Administration Manual' on the Linux OS.

In my local Guithub repository, thoses files have 777 has permissions. And I do not know how to fix this issue.

Do you have any idea ?

**Comment by lshep** [comment]
It looks like when it was cloned onto git.bioconductor.org it lost execute permissions.  I pushed up a change from a local checkout to add execute back and manually kicked off a build.  Lets see what the next report says.

**Comment by vjcitn** [comment]
make sure you remove any .Rproj files and bump version

**Comment by LiNk-NY** [comment]
Hi @samWieczorek 

Thank you for your submission. 
Please see the review below and respond to each item for the review to continue.

Best regards,
Marcel

---

# Pirat #3308

## DESCRIPTION

* Looks good.

## NAMESPACE

* Looks good.
* There seems to be a shiny app provided as part of the package. Consider
moving that functionality to a dedicated shiny package.

## vignettes

* Avoid using 1:N notation and use `seq_len` or `seq` instead in the
"transcriptomic integration" section.

* Minor: `which` is not needed for subsetting when a logical vector would work
the same, e.g., in "samples-wise correlations" section. If there are NAs, use
`na.rm = TRUE` with `colSums`, etc.
* Minor: Consider wrapping text to a max character limit of 80 columns. It is
easier to review and read.

## R

* Add a description to all fields in the data documentation (`@format` field
in `R/doc-data.R`)
* Remove or fix commented code in the `estimate_gamma()`,
`impute_block_llk_reset()`, `impute_block_llk_reset_PG()` functions, etc.
* It seems like `mcar` is not an override for `gamma_1` to be 0 as it is in a
multiple or `|` statement. Provide a more accurate description in the
documentation for `estimate_gamma`.
* Separate plots and calculation functions e.g., in `estimate_psi_df`
* Rename `install_Pirat` to more accurately reflect the intention of the
function.
* Move the shiny app to another package. Not all users may want to use
the shiny functionality.
* Use more descriptive argument names e.g., `ARG_VALUE_1` does not help the
user.
* Avoid using `cat` in functions that are not `print` or `show` methods, e.g.,
in `pipeline_llkimpute`.
* Provide the full list options in the argument as a character vector and use
`match.arg` in the body of the function, e.g., for `extension` argument in
`pipeline_llkimpute`, set it to `extension = c("a", "b", "c")`. This way the
user can see the full list of options in the usage section of the
documentation.
* Avoid using 1:N notation and use `seq_len` or `seq` for `for` loops; see
`get_indexes_embedded_prots`
* Consider moving plotting functions to another package. The package currently
has basilisk / python deps, shiny, and ggplot. This would render the package
fragile if any of those components begin to fail.
* Avoid nested for loops where possible; see `split_large_pg`.
* AFAIK, `basilisk` and `reticulate` operate differently yet you use them both.
Can you explain why both are needed?
* Resolve the following `R CMD check NOTE`:

```r
* checking R code for possible problems ... NOTE
estimate_gamma: no visible global function definition for ‘lm’
estimate_psi_df: no visible binding for global variable ‘var’
estimate_psi_df: no visible binding for global variable ‘x’
mod_Pirat_server : <anonymous>: no visible global function definition
  for ‘observe’
plot_pep_correlations: no visible global function definition for ‘cor’
plot_pep_correlations: no visible binding for global variable ‘group’
remove_NA_pep_reset: no visible global function definition for
  ‘remove_pep_from_idx’
Undefined global functions or variables:
  cor group lm observe remove_pep_from_idx var x
Consider adding
  importFrom("stats", "cor", "lm", "var")
to your NAMESPACE file.
```

## /

* Add `.github` folder to `.Rbuildignore`

## tests

* Increase coverage where needed:

```r
> covr::package_coverage(type = "all")
Pirat Coverage: 68.83%
R/mod_Pirat.R: 0.00%
R/reset_data.R: 0.00%
R/split_funcs.R: 48.28%
R/plots.R: 63.46%
R/pirat_utils.R: 82.29%
R/pipeline_impute.R: 92.78%
R/estimate_funcs.R: 93.88%
R/install.R: 100.00%
R/SE.R: 100.00%
R/utils.R: 100.00%
```

**Comment by samWieczorek** [comment]
Hi, 
I've corrected almost half of your comments. The rest are in the hands of a colleague who looks after certain parts of the code in Pirat.

I have 2 questions if I don't manage to fix everything by the time of release: what will the status of the package be?
- Is it accepted and remains in development?
- it's not put into development and its acceptance process is blocked?

About your question relative to the use of the reticulate package. I need it for two reasons:
- In the documentation of the package basilisk, there is an example on how to use their package an they explicitely use a function from the reticulate package to import Python libraries into the environment
-Pirat uses the  python package Torch but we have modified two files of this package. Theses files are in the inst/python directory. When I load the Python environment with functions of the package basilisk, I have to source those two customized Python scripts to overwrite the ones tha come with the library Torch. The function to do that does not exist in the basilisk package

**Comment by lshep** [comment]
If its accepted it will be added to the devel branch and will be  "released" in oct/nov bioc 3.20 release. You will still be able to reference the short url, have a bioconductor landing page,  and since devel still runs on 4.4 be available to users.

**Comment by LiNk-NY** [comment]
Hi @samWieczorek 
Thanks for following up. 
Please respond to the review points as there are a few items
that should be resolved before acceptance. 
It seems like the package will not be able to make it in time for the release. 
Best,
Marcel

**Comment by samWieczorek** [comment]
Hi Marcel, I think i did not understand where I have to answer the review points. I posted a message her : https://github.com/Bioconductor/Contributions/issues/3308 to explain that I have fixed many of those points. The last ones have been fixed Monday morning.

My last push on the github bioconductor repository has the status Ok

I will reply in the first email

Best

Sam


De : Marcel Ramos ***@***.***>
Envoyé : mardi 30 avril 2024 17:19
À : Bioconductor/Contributions ***@***.***>
Cc : WIECZOREK Samuel ***@***.***>; Mention ***@***.***>
Objet : Re: [Bioconductor/Contributions] Package Pirat (Issue #3308)


Hi @samWieczorek<https://github.com/samWieczorek>
Thanks for following up.
Please respond to the review points as there are a few items
that should be resolved before acceptance.
It seems like the package will not be able to make it in time for the release.
Best,
Marcel

**Comment by LiNk-NY** [comment]
> Hi Marcel, I think i did not understand where I have to answer the review points. I posted a message her : #3308 to explain that I have fixed many of those points. The last ones have been fixed Monday morning. My last push on the github bioconductor repository has the status Ok I will reply in the first email

Hi Samuel, @samWieczorek 
Please respond the _individual_ items in the review so that I may see 
which ones are still outstanding and take a look at the changes made so far. 

Thank you. 

-Marcel

**Comment by samWieczorek** [comment]
De : Marcel Ramos ***@***.***>
Envoyé : jeudi 25 avril 2024 20:47
À : Bioconductor/Contributions ***@***.***>
Cc : WIECZOREK Samuel ***@***.***>; Mention ***@***.***>
Objet : Re: [Bioconductor/Contributions] Package Pirat (Issue #3308)


Hi @samWieczorek<https://github.com/samWieczorek>

Thank you for your submission.
Please see the review below and respond to each item for the review to continue.

Best regards,
Marcel

________________________________
Pirat #3308<https://github.com/Bioconductor/Contributions/issues/3308>
DESCRIPTION

  *   Looks good.

NAMESPACE

  *   Looks good.
  *   There seems to be a shiny app provided as part of the package. Consider
moving that functionality to a dedicated shiny package. Has been removed

vignettes

·       Avoid using 1:N notation and use seq_len or seq instead in the
"transcriptomic integration" section. Fixed

·       Minor: which is not needed for subsetting when a logical vector would work
the same, e.g., in "samples-wise correlations" section. If there are NAs, use
na.rm = TRUE with colSums, etc. Fixed

·       Minor: Consider wrapping text to a max character limit of 80 columns. It is
easier to review and read. Fixed

R

  *   Add a description to all fields in the data documentation ***@***.*** field
in R/doc-data.R) Fixed
  *   Remove or fix commented code in the estimate_gamma(),
impute_block_llk_reset(), impute_block_llk_reset_PG() functions, etc. Fixed. The few remaining comments are only to better understand pieces of code
  *   It seems like mcar is not an override for gamma_1 to be 0 as it is in a
multiple or | statement. Provide a more accurate description in the
documentation for estimate_gamma. The manner we use mcar is an override because we use ‘|’ operator. Example : if mcar==TRUE then mcar | … | … <- TRUE
  *   Separate plots and calculation functions e.g., in estimate_psi_df.Removed
  *   Rename install_Pirat to more accurately reflect the intention of the
function. Fixed
  *   Move the shiny app to another package. Not all users may want to use
the shiny functionality. Fixed
  *   Use more descriptive argument names e.g., ARG_VALUE_1 does not help the
user. Fixed
  *   Avoid using cat in functions that are not print or show methods, e.g.,
in pipeline_llkimpute. Fixed
  *   Provide the full list options in the argument as a character vector and use
match.arg in the body of the function, e.g., for extension argument in
pipeline_llkimpute, set it to extension = c("a", "b", "c"). This way the
user can see the full list of options in the usage section of the
documentation. Fixed
  *   Avoid using 1:N notation and use seq_len or seq for for loops; see
get_indexes_embedded_prots Fixed
  *   Consider moving plotting functions to another package. The package currently
has basilisk / python deps, shiny, and ggplot. This would render the package
fragile if any of those components begin to fail. Shiny has been removed. Ggplot is mandatory because we need the plot generated with ggplot plot to control the algorithm quality
  *   Avoid nested for loops where possible; see split_large_pg. This is not possible in this case. Each PG which is too big is splitted in a different manner because they have different sizes and we must iterate on the new PGs to correctly complete them. Without a second loop, the code may be much more complex
  *   AFAIK, basilisk and reticulate operate differently yet you use them both.
Can you explain why both are needed?
I have removed calls to reticulate where it is possible (in the install_Pirat_env() function) and replaced by functions from basiliskpackage
However, I need the function import() from the reticulate package to import a little Python package stored in the inst/myModules directory. This is explained in the documentation and examples of the package basilisk. The function reticulate ::import() has no equivalent in the basilisk package

  *   Resolve the following R CMD check NOTE:

* checking R code for possible problems ... NOTE

estimate_gamma: no visible global function definition for ‘lm’

estimate_psi_df: no visible binding for global variable ‘var’

estimate_psi_df: no visible binding for global variable ‘x’

mod_Pirat_server : <anonymous>: no visible global function definition

  for ‘observe’ Fixed

plot_pep_correlations: no visible global function definition for ‘cor’

plot_pep_correlations: no visible binding for global variable ‘group’

remove_NA_pep_reset: no visible global function definition for

  ‘remove_pep_from_idx’ Fixed

Undefined global functions or variables:

  cor group lm observe remove_pep_from_idx var x

Consider adding

  importFrom("stats", "cor", "lm", "var")

to your NAMESPACE file. Only group remains as an Undefined global functions or variables but it is used in the ggplot2::aes() function.

  *   Add .github folder to .Rbuildignore Fixed

tests

  *   Increase coverage where needed:

> covr::package_coverage(type = "all")

Pirat Coverage: 68.83%

R/mod_Pirat.R: 0.00%

R/reset_data.R: 0.00%

R/split_funcs.R: 48.28%

R/plots.R: 63.46%

R/pirat_utils.R: 82.29%

R/pipeline_impute.R: 92.78%

R/estimate_funcs.R: 93.88%

R/install.R: 100.00%

R/SE.R: 100.00%

R/utils.R: 100.00%



New covr :
> covr::package_coverage(type = "all")
Pirat Coverage: 89.93%
R/pirat_utils.R: 83.16%
R/plots.R: 86.54%
R/estimate_funcs.R: 91.43%
R/pipeline_impute.R: 92.75%
R/install.R: 100.00%
R/SE.R: 100.00%
R/split_funcs.R: 100.00%





I hope i twill be sufficient

**Comment by LiNk-NY** [comment]
> * Avoid using cat in functions that are not print or show methods, e.g., in pipeline_llkimpute. Fixed 

> * Provide the full list options in the argument as a character vector and use match.arg in the body of the function, e.g., for extension argument in pipeline_llkimpute, set it to extension = c("a", "b", "c"). This way the user can see the full list of options in the usage section of the documentation. Fixed

I don't see where these were fixed. Any ideas? 

Note that there is a `StagedInstall` field in the `DESCRIPTION`. It may be removed in the future according to [WRE](https://cran.r-project.org/doc/manuals/R-exts.html#The-DESCRIPTION-file-1). Is there a reason why it should be set to "no"? 

Thanks, 
Marcel

**Comment by samWieczorek** [comment]
De : Marcel Ramos ***@***.***>
Envoyé : mardi 30 avril 2024 22:37
À : Bioconductor/Contributions ***@***.***>
Cc : WIECZOREK Samuel ***@***.***>; Mention ***@***.***>
Objet : Re: [Bioconductor/Contributions] Package Pirat (Issue #3308)


  *   Avoid using cat in functions that are not print or show methods, e.g., in pipeline_llkimpute. Fixed

>> A mistake. A last change of this file and I forgot to remove again the calls to cat() and print(). It is ok now


  *   Provide the full list options in the argument as a character vector and use match.arg in the body of the function, e.g., for extension argument in pipeline_llkimpute, set it to extension = c("a", "b", "c"). This way the user can see the full list of options in the usage section of the documentation. Fixed

>> Same reason as above (same file) is ok now


I don't see where these were fixed. Any ideas?

Note that there is a StagedInstall field in the DESCRIPTION. It may be removed in the future according to WRE<https://cran.r-project.org/doc/manuals/R-exts.html#The-DESCRIPTION-file-1>. Is there a reason why it should be set to "no"?

This is needed for the package basilik, and it is described in its documentation

StagedInstall: no should be set, to ensure that Python packages are installed with the correct hard-coded paths within the R package installation directory.

https://www.bioconductor.org/packages/release/bioc/vignettes/basilisk/inst/doc/motivation.html#for-package-developers



Many thanks



Sam

Thanks,
Marcel

**Comment by LiNk-NY** [comment]
Hi @samWieczorek 
Thank you for updating the package. 
The package has been accepted. 
Best regards,
Marcel

**Comment by samWieczorek** [comment]
Many thanks !

Sam

De : Marcel Ramos ***@***.***>
Envoyé : mercredi 1 mai 2024 16:20
À : Bioconductor/Contributions ***@***.***>
Cc : WIECZOREK Samuel ***@***.***>; Mention ***@***.***>
Objet : Re: [Bioconductor/Contributions] Package Pirat (Issue #3308)


Hi @samWieczorek<https://github.com/samWieczorek>
Thank you for updating the package.
The package has been accepted.
Best regards,
Marcel

**Comment by samWieczorek** [comment]
Hi

Thanks for the accepted my package Pirat.
Maybe I missed someting but I cannot see it in the packages of the release 3.19.
Is there something more I have to do ?

Best regards

Sam

De : bioc-issue-bot ***@***.***>
Envoyé : mercredi 1 mai 2024 16:20
À : Bioconductor/Contributions ***@***.***>
Cc : WIECZOREK Samuel ***@***.***>; Mention ***@***.***>
Objet : Re: [Bioconductor/Contributions] Package Pirat (Issue #3308)


Your package has been accepted. It will be added to the
Bioconductor nightly builds.

Thank you for contributing to Bioconductor!

Reviewers for Bioconductor packages are volunteers from the Bioconductor
community. If you are interested in becoming a Bioconductor package
reviewer, please see Reviewers Expectations<http://contributions.bioconductor.org/reviewer-resources-overview.html>.

**Comment by lshep** [comment]
I have not ingested new packages yet into the daily builds;  we were updating are DOI generation and had to wait for it to be completed.  The package since accepted after the release will be included in 3.20 but you can use the short link to advertise bioconductor.org/packages/Pirat as once it is in the system and landing pages generated this short link will always default to the release version of the package and while still in 3.20 the devel pages.   Since Bioconductor 3.19 and 3.20 users should still be able to utilize the package will minimal effort until 3.20 is officially released.

**Comment by lshep** [comment]
The default branch of your GitHub repository has been added to Bioconductor's
git repository as branch devel.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/samWieczorek.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("Pirat")`. The package 'landing page' will be created at

https://bioconductor.org/packages/Pirat

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/
